damnit/backend/db.py

```diff
@@ -194,18 +194,25 @@ def get_user_variables(self):
 
     def update_computed_variables(self, vars: dict):
         vars_in_db = {}
-        for row in self.conn.execute("""
-            SELECT name, title, description, attributes FROM variables
-            WHERE type == 'computed'
-        """):
-            var = dict(row)
-            vars_in_db[var.pop("name")] = var
-
-        updates = {n: v for (n, v) in vars.items()
-                   if v != vars_in_db.get(n, None)}
-        log.debug("Updating stored metadata for %d computed variables",
-                  len(updates))
         with self.conn:
+            # We want to read & write in the same transaction. This gets the
+            # write lock up front, to prevent deadlocks where two processes are
+            # both holding a read lock & waiting for a write lock.
+            self.conn.execute("BEGIN IMMEDIATE")
+            for row in self.conn.execute("""
+                SELECT name, title, type, description, attributes FROM variables
+                WHERE type == 'computed'
+            """):
+                var = dict(row)
+                vars_in_db[var.pop("name")] = var
+
+            updates = {n: v for (n, v) in vars.items()
+                       if v != vars_in_db.get(n, None)}
+            log.debug("Updating stored metadata for %d computed variables",
+                      len(updates))
+
+            # Write new & changed variables
+            # TODO: what if a new computed variable name matches an existing user var?
```

Comment:
Hmmm, I'm not sure about that. So far the validity of a context file is purely a function of the context file itself, which feels like a nice property, whereas this would mean its validity depends on the database too. That opens up possibilities like it's valid when you start executing it, but invalid by the time it finishes.

However we tackle this, though, I agree that it can wait for a separate PR.